bgpd/bgp_mplsvpn.c

```diff
@@ -1336,6 +1336,225 @@ leak_update(struct bgp *to_bgp, struct bgp_dest *bn,
+ * obtained
+ */
+static mpls_label_t bgp_mplsvpn_get_label_for_path_vrf(struct bgp_path_info *pi,
+						       struct bgp *to_bgp,
+						       struct bgp *from_bgp,
+						       afi_t afi)
+{
+	struct bgp_nexthop_cache *bnc = pi->nexthop;
+	struct bgp_label_per_nexthop_cache *blnc;
+	struct bgp_label_per_nexthop_cache_head *tree;
+	struct prefix *nh_to_use_ptr;
+	struct prefix nh_from_bnc = {0};
+
+	if (!bnc || !bnc->nexthop_num)
+		return from_bgp->vpn_policy[afi].tovpn_label;
+
+	/* check that the number of available nexthop is 1
+	 * ECMP is not supported
+	 * blackhole routes are not supported
+	 */
+	if (bnc->nexthop_num > 1 ||
+	    bnc->nexthop->type == NEXTHOP_TYPE_BLACKHOLE) {
+		/* if we are here, the bnc is valid, but does
+		 * not fit the needs
+		 * we will fallback to the default label
+		 */
+		bgp_mplsvpn_path_nh_label_unlink(pi);
+		return from_bgp->vpn_policy[afi].tovpn_label;
+	}
+
+	/* extract the resolved nexthop given by BNC
+	 * if nexthop type is not connected, then use the resolved nexthop
+	 */
+	if (bnc->nexthop->type != NEXTHOP_TYPE_IFINDEX) {
+		if (bnc->nexthop->type == NEXTHOP_TYPE_IPV4 ||
+		    bnc->nexthop->type == NEXTHOP_TYPE_IPV4_IFINDEX) {
+			nh_from_bnc.family = AF_INET;
+			nh_from_bnc.prefixlen = IPV4_MAX_BITLEN;
+			nh_from_bnc.u.prefix4 = bnc->nexthop->gate.ipv4;
+		} else {
+			nh_from_bnc.family = AF_INET6;
+			nh_from_bnc.prefixlen = IPV6_MAX_BITLEN;
+			memcpy(&nh_from_bnc.u.prefix6, &bnc->nexthop->gate.ipv6,
+			       sizeof(struct in6_addr));
+		}
+		nh_to_use_ptr = &nh_from_bnc;
+	} else
+		/* the nexthop is in bnc->prefix */
+		nh_to_use_ptr = &bnc->prefix;
+
+	/* at least one bnc is connected.
+	 * first, let us allocate a cache context
+	 */
+	tree = &from_bgp->mpls_labels_per_nexthop[family2afi(
+		nh_to_use_ptr->family)];
+	blnc = bgp_label_per_nexthop_find(tree, nh_to_use_ptr);
+	if (!blnc) {
+		blnc = bgp_label_per_nexthop_new(tree, nh_to_use_ptr);
+		blnc->to_bgp = to_bgp;
+	}
+
+	if (pi->label_nexthop_cache != blnc) {
+		/* Unlink from existing nexthop cache, if any.
+		 * This will also free the nexthop label cache entry
+		 */
+		bgp_mplsvpn_path_nh_label_unlink(pi);
+		if (blnc) {
+			/* updates NHT pi list reference */
+			LIST_INSERT_HEAD(&(blnc->paths), pi, label_nh_thread);
+			pi->label_nexthop_cache = blnc;
+			pi->label_nexthop_cache->path_count++;
+		}
+	}
+
+	if (blnc->label == MPLS_INVALID_LABEL)
+		bgp_lp_get(LP_TYPE_NEXTHOP, blnc,
+			   bgp_mplsvpn_get_label_per_nexthop_cb);
+	return blnc->label;
+}
+
+static mpls_label_t vpn_leak_from_vrf_get_label_for_path_vrf(
+	afi_t afi, safi_t safi, struct bgp_path_info *path_vrf,
+	struct bgp *from_bgp, struct bgp *to_bgp)
+{
+	mpls_label_t label_val;
+	bool attempt_policy_per_nexthop = false;
+	struct bgp *bgp_nexthop = NULL;
+	afi_t nh_afi;
+
+	/* only bgp updates that have a nexthop with a label can be sent */
+	if (CHECK_FLAG(from_bgp->vpn_policy[afi].flags,
+		       BGP_VPN_POLICY_TOVPN_LABEL_PER_NEXTHOP)) {
```

Comment:
No need to go further if per VRF mode is configured
```
if (!CHECK_FLAG(from_bgp->vpn_policy[afi].flags,
		       BGP_VPN_POLICY_TOVPN_LABEL_PER_NEXTHOP))
/* per label nexthop mode not configured. The per VRF mode is selected. 
 return the per VRF label*/
       return from_bgp->vpn_policy[afi].tovpn_label;
```

Reply:
done